### ISSUE TITLE
Provide method for looping over map elements

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#2955]
A looping construct to iterate over map elements would be useful:
```
BEGIN
{
  @x[1] = 100;
  @x[2] = 200;
  for ($kv : @x)
  {
    printf("key: %d\n", $kv.0);
    printf("val: %d\n", $kv.1);
  }
}
```

It would help with custom formatting when printing maps and I'm sure others will come up with more use-cases.

Related to #2807

**Comment by danobi** [comment]
Also similar to https://github.com/iovisor/bpftrace/issues/2384

**Comment by danobi** [comment]
Yes, this feature is often requested and would be nice. I recall discussing this at one point (maybe office hours?) - I think we mostly agreed that it would be good to transform the loop into a callback on behalf of the user. Introducing callback semantics to the language seems complex and somewhat unnecessary.

**Comment by viktormalik** [comment]
Yeah, just to summarize technical details - this should use the `bpf_for_each_map_elem` helper with a custom callback (created from the `for` loop body on behalf of the user), similarly to #2922.

I'll be happy to work on this but have other things in my queue atm so if someone wants to pick this up in the meantime, feel free to.

**Comment by jordalgo** [comment]
Sounds useful. The syntax also looks good. @viktormalik I'm also happy to pick it up if no one gets to it first (though similarly I have other things I need to get to before this).

**Comment by viktormalik** [comment]
Thanks @jordalgo. Just assign the issue to yourself once you start working on it (I'll do the same).

**Comment by danobi** [comment]
Whoever implements it ought to do a blog post. I think it's a pretty interesting topic (transforming a loop into a callback). Could help attract compiler enthusiasts when they see there's a bunch of low hanging stuff around :)

**Comment by jordalgo** [comment]
Just to note, we should probably test deletion/insertion into the same map during iteration to make sure there aren't weird side effects.

**Comment by thediveo** [comment]
is this released yet?

**Comment by jordalgo** [comment]
@thediveo Yes, it's in 0.21.0: https://github.com/bpftrace/bpftrace/releases/tag/v0.21.0

**Comment by ajor** [comment]
Closed by #3003
